### ISSUE TITLE
[Wishlist] Timeout for prefix key

### ISSUE THREAD

**Issue by bandresen** [noctuid/general.el#42]
I would like the following, if possible:

Using the leader "," I would like to use it in every every state (normal visual emacs insert)

When in a `:non-normal-prefix` I would like to have a short timeout where it will insert "," if I don't follow it up with a defined key.

**Comment by noctuid** [comment]
I've never liked/used keybinding systems that have timeouts, but I'll look into implementing this when I get the chance. 

Right now, you could kind of do this using `key-chord` by setting a longer temporary timeout, but then you're limited by the fact that key-chord only works with two characters. I'd probably add this feature by adding a timeout option to `general-key-dispatch`.

**Comment by noctuid** [comment]
One thing to note is that this wouldn't work well when emacs is lagging (`key-chord`
has this problem as well).

**Comment by bandresen** [comment]
If this is too much complexity and you don't think it's worth it then I just want to make it clear that it's not high priority for me either. It's just how I remember vim leader working.

For now I just use `, ,` and `, SPC` to insert `,` and `, ` respectively.

**Comment by noctuid** [comment]
All keybindings behave like that in vim if `timeout` is set. `general-simulate-keys` already implements 99% of the functionality needed for simulating this vim feature, so I might as well add a timeout option. 

I went ahead and implemented basic timeout functionality. I don't think it will have the same annoying issue as key-chord where you may accidentally trigger a command instead of a self insertion because of input lag, but there is still the potential for accuracy problems with the timeout is (and I don't know whether using `float-time` is the best way to go about implementing a timeout). Anyway, please test it and tell me if you find it usable. The timeout is in seconds and can be a floating point number. Here is an example definition:

```
(general-imap ","
              (general-key-dispatch #'self-insert-command
                :timeout 0.5
                "a" #'previous-line))
```
Here `,a` will run `previous-line` if `a` is pressed before the timeout. Otherwise it will input the characters `,a`. If you have a multikey sequence (e.g. "abc" instead of "a"), the timeout is reset after each key press (you'd have another 0.5 seconds starting at that time).

It seems to work okay, but let me know if this works for your use case or if you encounter any problems. If it ends up being usable, I'll keep it in and add documentation for it.

**Comment by bandresen** [comment]
I get `symbol's variable is void: timeout` on HEAD

I tried it both from my regular emacs setup (and carefully checked load order) and from an emacs -Q with only general.el/ included.

**Comment by noctuid** [comment]
Oops, my bad. I missed a line when committing. Try it now.

**Comment by bandresen** [comment]
How would the syntax for multiple keybindings be?

```
    (general-imap ","
                  (general-key-dispatch #'self-insert-command
                    :timeout 0.4
                    "i" #'ibuffer))
    
    (general-imap ","
                  (general-key-dispatch #'self-insert-command
                    :timeout 0.2
                    "g" #'magit-status))
```
That reminds everything but the latest (,g) while
```
    (general-imap ","
                  (general-key-dispatch #'self-insert-command
                    :timeout 0.4
                    "i" #'ibuffer)
                  (general-key-dispatch #'self-insert-command
                    :timeout 0.2
                    "g" #'magit-status))
```
Results in the following:
`Debugger entered--Lisp error: (wrong-type-argument arrayp general-dispatch-self-insert-command)`

**Comment by bandresen** [comment]
Also, this is how my `general-define-key` partially looks currently:
```
     "g"   '(:ignore t :which-key "magit")
     "gs"  'magit-status
     "gd"  'magit-diff
     "gc"  'magit-commit
     "gl"  'magit-log
     "gp"  'magit-push
     "gw"  'magit-stage-file ;; "write"
     "gr"  'magit-unstage-file ;; "remove"
     "gg"  'counsel-ag
```

Is that supposed to work?

(And let me repeat, if this is too much work I can deal with losing , , and , SPACE.

**Comment by noctuid** [comment]
You can't have different timeouts under the same key currently (for the same timeout, you just add the keys in the `general-key-dispatch`. I don't think having different timeouts is possible in vim by default (feel free to correct me). Would you find that useful? It seems like it would be confusing in most cases. The error you got is because you have two commands after a key sequence (you need one command for every key sequence). For the same timeout, the command would be:
```
(general-imap ","
              (general-key-dispatch #'self-insert-command
                :timeout 0.4
                "i" #'ibuffer
                "g" #'magit-status))
```

As for your second example, are you encountering some problem with that? It seems fine to me (though I don't know the full command so you could be binding under a non-prefix key).

**Comment by bandresen** [comment]
I didn't want different timeouts, I was just confused about the syntax.

```
    (general-imap ","
                  (general-key-dispatch #'self-insert-command
                    :timeout 0.4
                    ";" '((lambda () (interactive)
                            (ansi-term (getenv "SHELL")))
                          :which-key "ansi-term")
                    "i" #'ibuffer
                    "g" #'magit-status))
```

when I do `, ;` then I get: 
`Debugger entered--Lisp error: (wrong-type-argument commandp ((lambda nil (interactive) (ansi-term (getenv "SHELL"))) :which-key "ansi-term"))`

When I try:
```
    (general-imap ","
                  (general-key-dispatch #'self-insert-command
                    :timeout 0.4
                    "d"  '(:ignore t :which-key "describe")
                    "dk" 'describe-key))
```

then I get `Key sequence d k starts with non-prefix key d`

**Comment by noctuid** [comment]
`general-key-dispatch` doesn't currently support extended definitions. I could potentially at support for them, but for this case at least (`:which-key` and `:timeout`), it doesn't seem like it would be too useful, since the timeout would probably occur before you could read a which-key popup. If you want support for this, I'll open a separate issue.

**Comment by bandresen** [comment]
I absolutely agree that it doesn't make much sense for `:which-key` and `:timeout` simultaneously.

I just would like to define my leader keybindings only once and then have it work in every state. Emacs and Insert should just be with a timeout.

**Comment by noctuid** [comment]
That would be a little difficult to do. I don't think I want to parse all the mappings and find the ones that need to be rewritten in to a `general-key-dispatch`. Maybe what I could do to prevent the need to duplicate all your leader keybindings is allow `general-key-dispatch` to get its keybindings from an existing keymap. For example, you'd bind keys in normal state as you would normally, but you would explicitly name the prefix keymap for `,` to something like `my-leader-map`. Then you could bind `,` in emacs and insert state with the same keybindings using something like this:

```
(general-iemap ","
               (general-key-dispatch #'self-insert-command
                 :timeout 0.4
                 :copy-keymap my-leader-map))
```

Would that be better?

**Comment by bandresen** [comment]
That would be great!

**Comment by noctuid** [comment]
I've added it as `:inherit-keymap`. It will work even if you bind more `,<keys>` in normal state afterwards. Here's an example:

```
(general-nmap :prefix ","
              :prefix-command 'my-prefix-map
              "g" #'magit-status)

(general-imap ","
              (general-key-dispatch #'self-insert-command
                :timeout 0.25
                :inherit-keymap my-prefix-map))
```

Let me know if you encounter any problems.

**Comment by bandresen** [comment]
I love it!

First impressions is that it does everything I want without configuration duplication. Thank you very much!

**Comment by noctuid** [comment]
Great! If you have any problems, I'll reopen the issue.
